design-proposals/klip-33-key-format.md

```diff
@@ -0,0 +1,299 @@
+    * Automatic repartitioning of streams and tables for joins where key formats do not match: such
+      joins will result in an error initially.
+  
+    Included in this milestone:
+    
+    * Addition of a new optional `KEY_FORMAT` property in the `WITH` clause, to set the key format.
+    * Addition of a new optional `FORMAT` property in the `WITH` clause, to set both the key & value formats.
+    * Addition of new server configuration to provide defaults for key and value formats.
+    * Removal of requirement for `VALUE_FORMAT` property in the `WITH` clause of CT/CS statements, where
+      the server configuration provides a default.
+    * Support for additional key column data types, as JSON supports them:
+        * `DECIMAL`
+        * `BOOLEAN`
+    * Full support of these key formats for all supported SQL syntax  
+1. **Auto-repartitioning on key format mismatch**. Adds support for automatic repartitioning of 
+   streams and tables for joins where key formats do not match.
+1. **Schema Registry support**: Adds support for reading and writing schemas to and from the schema
+   registry.
+1. **JSON_SR support** Adds support for the `AVRO` key format, inc. schema registry integration.
+1. **Avro support** Adds support for the `AVRO` key format, inc. schema registry integration.
+1. **Delimited support**: Adds support for the `DELIMITED` key format.
+1. **Optional KAFKA_TOPIC property**: makes the `KAFKA_TOPIC` property optional everywhere and add
+   configuration to control case and case sensitivity of topic names.
+1. **Blog post**: write a blog post about the new features. (Potentially more than once if 
+  work span multiple releases).
+   
+LOE TBD once scope and design confirmed.
+
+## Documentation Updates
+
+New server config and new `CREATE` properties will be added to main docs site.
+
+There are no incompatible changes within the proposal, so no demos and examples _must_ change.
+However, it probably pays to update some to highlight the new features. We propose updating the 
+Kafka micro site examples to leverage the new functionality, as these have automated testing.  
+It may be worth changing the ksqlDB quickstart too - TBD, as this will require extending DataGen 
+to support other key formats. Something we may want in scope anyway - or should be end-of-life 
+DataGen in favour of the datagen connector?
+
+## Compatibility Implications
+
+As mentioned above, existing query plays already include key formats for all topics. So existing
+queries will continue to work.
+
+Without `ksql.persistence.default.format.key` set to `KAFKA` existing queries in the form:
+
+```sql
+CREATE TABLE USERS (
+    ID BIGINT PRIMARY KEY,
+    NAME STRING
+  ) WITH (
+    KAFKA_TOPIC='USERS',
+    VALUE_FORMAT='JSON'
+  );
+```
+
+...will start failing, as they do not specify the `KEY_FORMAT`. We therefore propose the server 
+config shipped with ccloud and on-prem releases has `ksql.persistence.default.format.key` set to 
+`KAFKA`.
```

Comment:
+1 as above, why not have this be the in-code default?

Reply:
I see two main ways of implementing this:
 - default in the `KsqlConfig`
 - defaults in the server properties files.

The former has the pro of handling both on-prem and ccloud, but the con that it might not be possible to unset the property, i.e. to add config to say 'I don't want a default'.  Not sure that's an issue though.